### ISSUE TITLE
max_in_flight for Input:Kafka and output:Sql

### ISSUE THREAD

**Issue by mangatmodi** [Jeffail/benthos#651]
So we have batching at the Kafka input

```
    batching:
      count: 10
      period: 100ms
```

And then we have no batching at the SQL output(It is experimental If I am right). 

_I was wondering how will `max_in_flight` behave in such a configuration?_

1. If we send x messages to SQL (max_in_flight: x) then if one of them fails, do we fail all inserts to the DB. Do all in-flight messages share the same transaction?
2. If not, then if let's say a message m1 fails and m2 is a success but m2 has a higher offset in Kafka, how will we commit the offsets in kafka? 
3. What about the ordering inside the max_in_flight?

Basically, we want to write into the output in the same order for a partition as it was in Kafka. We are doing an upsert and hence that is required. How can we optimize for performance?

**Comment by Jeffail** [comment]
Hey @mangatmodi, if you're batching messages at the input level then you don't need to configure batching at the output level, messages batched by your Kafka consumer will be sent within the same transaction by the SQL output, which means ordering is preserved.

When you set `max_in_flight` at the output level those transactions will not change, but it'll allow the output to execute multiple transactions at the same time.

Since the Kafka input batches messages at the topic/partition level it means you can increase the number of max in flight transactions and benefit from increased throughput, and ordering is still guaranteed at the topic/partition level.

If you have a config something like this:

```yaml
input:
  kafka:
    addresses:
      - TODO:9092
    topics:
      - TODO
    consumer_group: foogroup
    client_id: benthos_kafka_input
    batching:
      count: 10
      period: 100ms
    
output:
  sql:
    driver: mysql
    data_source_name: TODO
    query: TODO
    args: [] # TODO
    max_in_flight: 10
```

And say you have 5 partitions in the topic `TODO`, then it would mean you'd be inserting data into to your table with transactions of a maximum 10 rows each, and up to 5 transactions in flight at a given time as this will be limited by the number of partitions. Messages of each partition will always be inserted in order as the transaction must succeed before subsequent offsets will be pushed.

Then if you were to add more partitions to your topic, or consume from more topics, you would potentially start hitting the cap of 10 max in flight transactions.

**Comment by mangatmodi** [comment]
@Jeffail 
Much appreciated. It works well for us. 

Thanks :)

**Comment by mangatmodi** [comment]
So we tried this got 7.5 times more throughput :)
